### PR TITLE
plumbing/object: Support mergetag in merge commits

### PR DESCRIPTION
**Pull request by adityasaky**
When a merge commit is created from merging a signed tag, the tag object is embedded in the commit object. This commit adds support for this tag data when encoding and decoding a commit object.

It might be possible to embed the Tag struct in the Commit struct. However, that will also require additions to the tag object to use reader / writer, like the Signature object currently does.

NOTE: I haven't included any tests as I'm currently unfamiliar with the test workflow used in this repo. Some pointers would be appreciated!

### CODE REVIEW

**Comment by adityasaky** [comment]
Here's an example of a commit that includes the mergetag field: https://github.com/torvalds/linux/commit/cac405a3bfa21a6e17089ae2f355f34594bfb543.

```
[linux] $ git cat-file -p cac405a3bfa21a6e17089ae2f355f34594bfb543
tree d2d33b5eea7df14b5c2e5cd49304f957837e3321
parent 50768a425b46ad7d98f6d88c22d41aa026c463cf
parent b4c639f699349880b7918b861e1bd360442ec450
author Linus Torvalds <torvalds@linux-foundation.org> 1695746648 -0700
committer Linus Torvalds <torvalds@linux-foundation.org> 1695746648 -0700
mergetag object b4c639f699349880b7918b861e1bd360442ec450
 type commit
 tag for-6.6-rc3-tag
 tagger David Sterba <dsterba@suse.com> 1695661654 +0200
 
 for-6.6-rc3-tag
 -----BEGIN PGP SIGNATURE-----
 
 iQIzBAABCgAdFiEE8rQSAMVO+zA4DBdWxWXV+ddtWDsFAmURvloACgkQxWXV+ddt
 WDt+CQ/+NgBtQn7eyABsdHzXWPxpFyGZrdw5ldKnly3G+WDW2GKMaZ6CpDuEZGNQ
 vMAkSGX5LIHXvO79pDnGG0i+bRINWrc5HZVZ/p5Da6wplBTgIPlbLmxaZX9MJLbx
 j7Oz37GXiQJY8BxnVCnsb+bhhTrTbO9HFUQr/nxefIvu22OBdL1WXYcfuBOeEsFG
 qr/aeC52YqCVgXvt+8a5DqAKE0NWc4PFMFUMo4vlf1xuL652fvff7xiup1CAIgBh
 qsCa17E7q+qjri2phAhbFNadfpH5wGfyjTWScOlaFuXjRhW2v2oqz3WU5IQj4dmu
 PI+k++PLUzIxT0IcjD1YbZzRFaEI6fR2W0GA4LK08fjVehh2ao5jOjtRgLl8HlqG
 qC5fslAPzUxRmwMmCjSGfXF14sgtyLy8eVWf69xn06/1cbEmfHDrWNXP1QHuq6eT
 Jqy8Ywia3jRzzfZ1utABJPLBW4hFQKkyobtyd67fxslUFmtuLvLqGTiOdmVFiD9K
 o+BF2xjEz2n8O1+aRZk5SFNC9zcaASaRg/wQrhvSI9qxM18fh4TXgKQOniLzAK7v
 lZc+JkegFW4CVquCUpmbsdZAOpVNRXfPOJIt/w6G+oRbaiTvPUnrH+uyq8IGREbw
 E7d8XIP0qlF0DQBGK4Mw/riZz/e5MmEKNjza6M+fj2uglpfWTv4=
 =6WEW
 -----END PGP SIGNATURE-----

Merge tag 'for-6.6-rc3-tag' of git://git.kernel.org/pub/scm/linux/kernel/git/kdave/linux

Pull btrfs fixes from David Sterba:

 - delayed refs fixes:
     - fix race when refilling delayed refs block reserve
     - prevent transaction block reserve underflow when starting
       transaction
     - error message and value adjustments

 - fix build warnings with CONFIG_CC_OPTIMIZE_FOR_SIZE and
   -Wmaybe-uninitialized

 - fix for smatch report where uninitialized data from invalid extent
   buffer range could be returned to the caller

 - fix numeric overflow in statfs when calculating lower threshold
   for a full filesystem

* tag 'for-6.6-rc3-tag' of git://git.kernel.org/pub/scm/linux/kernel/git/kdave/linux:
  btrfs: initialize start_slot in btrfs_log_prealloc_extents
  btrfs: make sure to initialize start and len in find_free_dev_extent
  btrfs: reset destination buffer when read_extent_buffer() gets invalid range
  btrfs: properly report 0 avail for very full file systems
  btrfs: log message if extent item not found when running delayed extent op
  btrfs: remove redundant BUG_ON() from __btrfs_inc_extent_ref()
  btrfs: return -EUCLEAN for delayed tree ref with a ref count not equals to 1
  btrfs: prevent transaction block reserve underflow when starting transaction
  btrfs: fix race when refilling delayed refs block reserve
```

I also created a version where the merge commit is also signed.

```
[test-signed-tag-merge] $ git cat-file -p HEAD
tree dcc0f810d1e7315d4c8e9c1f6fae99e8b161ae25
parent 58eb27c1bd5696ab45b9cdc054d7e12feca003fb
parent 7b6ed05bc0895f7453564f12f2f719f3a3b52031
author Aditya Sirish <aditya@saky.in> 1695827930 -0400
committer Aditya Sirish <aditya@saky.in> 1695827930 -0400
mergetag object 7b6ed05bc0895f7453564f12f2f719f3a3b52031
 type commit
 tag change
 tagger Aditya Sirish <aditya@saky.in> 1695827841 -0400

 change
 -----BEGIN PGP SIGNATURE-----

 iQEzBAABCAAdFiEE4ylBKZy4wNk9zyesuDEQ0BJUVgQFAmUUR4EACgkQuDEQ0BJU
 VgSZaAgAwn3StCshXDozuq7bZj4R8T+HjEDwpjqASoq5mS2rTo4sweTDpXvzBTWX
 IzugCHlpdAEhImj35VzJLpi2FPiVrhwLS5N/gCAIUpdPvu1oqNN7e6v9P6UOTBIM
 vcgRm04UR3H2Ov2cCTji8URnivBAToHnd1ucosexS+VkiViK5xtiS403kDLFPWbG
 CraB3s+4MRhSA7XnBo2Oc8Mg+9c7p7hjjHg6Xmt34yl8/Gr/QwOWhBj9KZJAYpM/
 3ech1fvpkagSjf8MnK4LGartCnRuidT2r5HKBxsEL3PPghuqmSvaAGDSoQVRpPlT
 ug53KhQhMpi0Wo1ibkiovYT+iNNbJQ==
 =0jbQ
 -----END PGP SIGNATURE-----
gpgsig -----BEGIN PGP SIGNATURE-----

 iQEzBAABCAAdFiEE4ylBKZy4wNk9zyesuDEQ0BJUVgQFAmUUR+AACgkQuDEQ0BJU
 VgQDXQgAh9IDVPPmDLnLxtZ2YnIVwbrL2oj73mKlt84mUXij/TPVpZ8D7V5NUaSG
 OTZ9amKd+iBIpDsL5UJJj5Fnm24sybNZrbfDaXCCvoBgSb2CFySALPuVMNkfcY/C
 FxWKUcGxkNqdYB9X7g/U9YB/7+hsKwuPgaOf7abLXiJsftjmvWSIt9dPSsoQr3GM
 OBIbpRkhERK1AhyREwBlTPnn0k7MivaLZjoQ9c8GyBJPP3cGe+9kUmbZCGAlj9Kv
 Rxu2SCv+n7ReEAm/E3HFkQHrhzwHbs82B0m+yDWiYI23EQeMdPFjBx2hABxty+eS
 LvTjgw4FelsMndjtw+dmucn7Py8P7A==
 =vydX
 -----END PGP SIGNATURE-----

Merge tag 'change'

change

**Comment by pjbgf** [comment]
@adityasaky nice PR!

The majority of our tests use [gocheck](https://labix.org/gocheck). But feel free to go standard Golang if that is more convenient as we may move on from `gocheck` as it is no longer maintained.

[plumbing/object/commit_test.go](https://github.com/go-git/go-git/blob/809f9df1b76258a311a20c76d346e86aca0a08f8/plumbing/object/commit_test.go) has some examples for signature related tests. 

If you want to do state based tests, you could also add a new fixture at [go-git/go-git-fixtures](https://github.com/go-git/go-git-fixtures) and then [fetch it by tag](https://github.com/go-git/go-git/blob/809f9df1b76258a311a20c76d346e86aca0a08f8/plumbing/format/packfile/packfile_test.go#L121) for example but that is not a hard requirement.

File: plumbing/object/commit.go [hunk 1]
```diff
@@ -286,6 +304,24 @@ func (c *Commit) encode(o plumbing.EncodedObject, includeSig bool) (err error) {
...
+		if _, err = fmt.Fprint(w, "\n"+headermergetag+" "); err != nil {
+			return err
+		}
+
+		// Split tag information lines and re-write with a left padding and
+		// newline. Use join for this so it's clear that a newline should not be
+		// added after this section. The newline will be added either as part of
+		// the PGP signature or the commit message.
+		// FIXME: should this and PGP signature be moved into a generic
+		// multi-line field handler?
```
Comment (pjbgf) [inline plumbing/object/commit.go:317]: We can go with this approach and then refactor it as part of https://github.com/go-git/go-git/pull/705.

**Comment by adityasaky** [comment]
@pjbgf I added a simple test with a sample tag object's representation. Let me know what you think!

File: plumbing/object/commit.go [hunk 2]
```diff
@@ -17,9 +17,10 @@ import (
 )
 
 const (
-	beginpgp  string = "-----BEGIN PGP SIGNATURE-----"
-	endpgp    string = "-----END PGP SIGNATURE-----"
-	headerpgp string = "gpgsig"
+	beginpgp       string = "-----BEGIN PGP SIGNATURE-----"
```
Comment (pjbgf) [inline plumbing/object/commit.go:20]: We could add a comment and link to the gitformat docs:

https://github.com/git/git/blob/bcb6cae2966cc407ca1afc77413b3ef11103c175/Documentation/gitformat-signature.txt#L153

**Comment by pjbgf** [comment]
@adityasaky please squash the commits.

File: plumbing/object/commit.go (hunk 2 above)
Comment (adityasaky) [reply plumbing/object/commit.go]: Done!

**Comment by adityasaky** [comment]
Added a link to the signature format and cleaned up commit history.

**Review by pjbgf — APPROVED** [review]
@adityasaky great job, thank you for working on this. :bow:

**Comment by adityasaky** [comment]
Thanks for the quick merge!

**Pull request merged**